### PR TITLE
PERF: join empty frame

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Similar to #45838, but for `DataFrame.join`. 

`DataFrame.join` already had a fast path for joining with an empty frame but it only covered a limited set of cases. This PR makes the fast path a bit faster and covers additional cases. 

```
import pandas as pd 
import numpy as np

N = 10_000_000

df = pd.DataFrame({'A': np.arange(N)})
df_empty = pd.DataFrame(columns=['B', 'C'], dtype='int64')

%timeit df.join(df_empty, how='inner')
932 ms ± 83.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)       <- main
285 µs ± 4.29 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)  <- PR

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -4544,12 +4544,22 @@ def join(
 
         if len(other) == 0 and how in ("left", "outer"):
```
Comment (jreback) [inline pandas/core/indexes/base.py:4545]: can you make a `if len(other)` and if `len(self)` clause then sub if's here for the cases

**Review by jreback — CHANGES_REQUESTED** [review]
do we have sufficient asv's to cover this?

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (lukemanley) [reply pandas/core/indexes/base.py]: Done.

**Review by jreback — APPROVED** [review]
lgtm ping on green

**Comment by lukemanley** [comment]
@jreback - greenish, errors look unrelated

**Comment by mroeschke** [comment]
Could you merge in main one more time? (failures look unrelated but good to be sure)

**Comment by lukemanley** [comment]
@mroeschke - merged main. greenish again, let me know if you think the error is related

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lukemanley. The failure was unrelated

**Pull request merged**